### PR TITLE
perf(k8sClient): use a cache-based version of k8s client

### PR DESCRIPTION
**Pull request by fpetkovski**
The current metacontrollers use a plain controller-runtime client when talking to the API server.
As a result, each read and list action goes directly to the server instead of using a cache.

This commmit initializes the client to be the same as the one from the controller manager.
In this way, reading actions will always be backed by a cache, while write actions
will be executed directly against the API server.

Signed-off-by: fpetkovski <filip.petkovsky@gmail.com>

### CODE REVIEW

**Comment by grzesuav** [comment]
hi @fpetkovski , thank you for contribution, I missed the notification email about this one.

The one thing I have doubts here is redundant assignment in context creation method - https://github.com/metacontroller/metacontroller/blob/3fef5b4cabeb158a248bbd4ce7effe5830eef006/pkg/controller/common/common.go#L96

I was thinking how to handle it better, for sure mentioned code should be removed, I just thinking what is the best way to provide client there :
* assign as it is currently
* make `k8sClient` a parameter in `Start` method and pass it there - https://github.com/metacontroller/metacontroller/blob/3fef5b4cabeb158a248bbd4ce7effe5830eef006/pkg/server/server.go#L100



what do you think ?

**Review by grzesuav — CHANGES_REQUESTED** [review]
As in comment above.

**Comment by fpetkovski** [comment]
Hi @grzesuav, yes, that is a good point. I also did not see any good way to do this because there seems to be a circular dependency between the manager (due to broadcaster) and the execution context (due to the client from manager). 

I would lean towards simply removing https://github.com/metacontroller/metacontroller/blob/3fef5b4cabeb158a248bbd4ce7effe5830eef006/pkg/controller/common/common.go#L96

My concern with passing the client in https://github.com/metacontroller/metacontroller/blob/3fef5b4cabeb158a248bbd4ce7effe5830eef006/pkg/server/server.go#L100 is that the entire initialization logic starts to get a bit convoluted and hard to follow.

Wdyt?

**Comment by grzesuav** [comment]
@fpetkovski agreed :+1:

**Review by grzesuav — APPROVED** [review]

**Comment by grzesuav** [comment]
thanks @fpetkovski ! Sorry for delay, got busy weekend

**Comment by fpetkovski** [comment]
@grzesuav no worries, was a long weekend for me as well :)
